### ISSUE TITLE
LG-60: Make Liberty JSON log field names be configurable

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#6079]
AHA: https://bigblue.aha.io/features/LG-60

In some cases it's helpful to be able to change the field names used in Liberty's JSON log output.  For example, I may want to rename the "message" field to "log" to be consistent with records Filebeat creates when I'm running in ICP.

- [x] check to see if changing liberty field name from `message` to `log` can make icp replace existing log field with current message field contents
- [ ] determine best set of field names to use in ICP to maximize consistency (in conjunction with other products)

could possibly use some sort of mapping table in Liberty to do this:
`WLP_LOGGING_MESSAGE_FIELDS="message:log, ibm_datetime:@timestamp, loglevel:level"`

ID issue: https://github.ibm.com/was-liberty/liberty-docs/issues/1563

**Comment by donbourne** [comment]
this may be safer to do in Logstash than in each product (since it will be difficult for every product to provide this mapping capability). 

As an example of why this is hard, log4J has a JSONLayout that writes JSON with the message text in a "message" field.  If Liberty were to write its own messages with a different JSON field name it wouldn't help bring log4J JSON into consistency.

**Comment by kbaillie** [comment]
When the constant in LogFieldConstants is changed from "message" to "log", the field name in the log is changed. When the logs are seen in icp, the existing log field is replaced with current message field contents.

**Comment by donbourne** [comment]
Reopening this issue as goal is to provide solution for this in OL

**Comment by donbourne** [comment]
see also comment from SO related to using Liberty with LogDNA in IBM Cloud at https://stackoverflow.com/questions/55189309/how-to-tweak-the-json-log-format-in-websphere-liberty/55190516#55190516

**Comment by fmhwong** [comment]
Target 19.0.0.8
NLS 7/19
DCUT 7/24

**Comment by donbourne** [comment]
need to leave issue open to get approvals

**Comment by fmhwong** [comment]
Target 19.0.0.9
Code completion 8/21/2019

**Comment by fmhwong** [comment]
## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

**Comment by fmhwong** [comment]
@gscottj This feature doesn't have UI.  May I have accessibility approval?
@gecock SVT is not required for this feature.  May I have the SVT approval?
@chirp1 ID content has been provided in the separate ID issue.  May I have ID approval?
@jhanders34 We don't expect performance hit in this feature.  Please assess.

**Comment by gscottj** [comment]
This feature does not have an user interfaces, therefore there is no impact to accessibility compliance.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi, looking at https://bigblue.aha.io/features/LG-60 it does not appear that the WAD Review has been completed (though it looks more likely the activity simply hasn't been updated). Please can you add the "WAD Approval" label to this epic once the WAD has definitely been reviewed and approved, thanks!

**Comment by fmhwong** [comment]
@dave-waddling Ian approved the WAD on 8/13.  I just added the WAS approval label.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @fmhwong thanks for the quick update. I've now completed the review of the Feature Test Summary and, all looks good to me. If the testing passes cleanly in the up-coming SOE I'll be able to grant a FAT Focal Approval early next week.

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same BOX folder.  Thanks!

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi, all testing passed cleanly in the latest SOE, so as per my previous comment I'm granting the FAT Focal Approval at this time.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by skasund** [comment]
Providing STE approval early. Please provide the slides as soon as possible. Thanks!

**Comment by skasund** [comment]
@Yushan-Lin Thanks for the STE slides

**Comment by marikaj123** [comment]
@Yushan-Lin - Does the feature require translation?

**Comment by Yushan-Lin** [comment]
@marikaj123 yes translation is needed

**Comment by fmhwong** [comment]
This file contains new messages.  https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.logging/resources/com/ibm/ws/logging/internal/resources/LoggingMessages.nlsprops and I just checked the translated files were merged to integration yesterday.

**Comment by chirp1** [comment]
David Mueller is working on the doc updates through this ID issue: https://github.com/OpenLiberty/open-liberty/issues/6079 . I'm approving this epic.

**Comment by fmhwong** [comment]
After discussion with Alasdair, he recommended to make some changes.  This will not be GA in 19.0.0.9.  We just tagged this as beta and plan to GA it in 19.0.0.10.

**Comment by fmhwong** [comment]
@marikaj123 May I have globalization approval?

**Comment by fmhwong** [comment]
@donbourne May I have serviceability approval?

**Comment by donbourne** [comment]
@fmhwong can you please fill in the answers to the questions above in Serviceability Approval comment?  Great that you've got approval from SVT team, but other questions are important too.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @fmhwong, in your comment 19 days ago you mentioned making some additional changes requested by Alasdair. Please can you:
-  Add issues to this epic under which those changes are being delivered;
- Remove the 'FAT complete' label until all code/testing is delivered for this feature;
- Update the Feature Test Summary (#8643) to explain how the changes are being tested.

In the meantime I'm going to remove the FAT Focal Approval to ensure we do our due diligence once this feature is complete.

**Comment by Yushan-Lin** [comment]
Changes to FAT: https://github.com/OpenLiberty/open-liberty/issues/8813
PR: https://github.com/OpenLiberty/open-liberty/pull/8823

**Comment by Yushan-Lin** [comment]
**1. WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?**
Yes, the WAD addresses problems customers might encounter. Warning messages were implemented and provides information about issues customers might encounter.
2. Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
 a) What problem paths were tested and demonstrated?

- User provides a defaultFieldName that does not exist

- User provides a source that does not exist

- User provides too many tokens in the jsonFieldMappings format

- User provides too little tokens in the jsonFieldMappings format

b) Who did you demo to? SVT 
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
3. SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered. a) Who conducted SVT tests for this feature? b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
4. Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

**Comment by fmhwong** [comment]
@dave-waddling We have updated the feature test summary.  Please review.

**Comment by fmhwong** [comment]
@sabolo Hi Kevin, Dave is on vacation.  Can you help on the FAT approval?

**Comment by sabolo** [comment]
Hi @fmhwong :  Could you confirm that the only change that was requested for this feature was to change JSONFields to JSONFieldMappings?    The original comment we had implied multiple changes were requested so looking for confirmation that you have completed all the required changes for this feature.  Assuming there were no other items to be done I will complete the sign-off on confirmation.  If additional changes, please can you detail what they were and whether they are now complete.  Many thanks

**Comment by fmhwong** [comment]
Hi @sabolo, This feature introduces the capability of renaming JSON fields in logs when JSON logging is enabled.  @dave-waddling gave FAT approval on 8/22 for this feature.  Alasdair made a comment about the attribute name in the server configuration.  Therefore, we changed the `jsonFields` to `jsonFieldMappings` and modified the FAT accordingly.  Dave wanted to re-assess after the change but he is out on vacation.  To summarize, the net new change after Dave's original approval is just the rename of the attribute name.

**Comment by marikaj123** [comment]
@fmhwong - Please close the feature since all the approvals were granted and the feature made it into the GM build.

**Comment by fmhwong** [comment]
Blog https://github.com/OpenLiberty/open-liberty/issues/9087

**Comment by fmhwong** [comment]
Closing
